### DISCUSSION TITLE
`SConstruct` not recognizing that a piped action requires a certain executable to run

### DISCUSSION THREAD

**Discussion by awa5114** [SCons/scons#4354] [Q&A]
I have a directory with the following contents:

```
|   file.in
|   SConstruct.py
|   
\---modify
        modify.cpp
        SConscript.py
```

Here are the contents of each file:

# `file.in` 

```
in file
```

# `SConstruct.py`

```python
SConscript(["modify/SConscript.py"])


Command(
    "file.out",
    "file.in",
    action = [
        Copy("tempfile", "$SOURCE"),
        ".\modify.exe tempfile appendtext",
        Copy("$TARGET", "tempfile")
    ]
)
```

# `modify/modify.cpp`

```cpp
#include <iostream>
#include <fstream>
#include <filesystem>

int main(int argc, char* argv[])
{
    std::filesystem::path currentPath = std::filesystem::current_path();
    std::filesystem::path fullPath = currentPath / argv[1];
   
    if (!std::filesystem::exists(fullPath))
    {
        std::cout << "File doesn't exist, aborting" << std::endl;
        return 1;
    }

    std::ofstream outfile;
    outfile.open(fullPath.string(), std::ios_base::app);

    outfile << argv[2] << std::endl;

    return 0;
}
```

# `modify/SConscript.py`

```python
env = Environment(tools=["mingw"])

env.Program("modify.cpp")
Command("../modify.exe", "modify.exe", Copy("$TARGET", "$SOURCE"))
```

The idea here is to compile `modify.cpp` to `modify.exe`, copy `modify.exe` to the root folder and then run it against `file.in` to generate a new file. However when I run `scons` the `SConstruct` executes the `Copy` command first, and then naturally runs into a problem because `modify.cpp` has not been compiled yet:

```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Copy("tempfile", "file.in")
.\modify.exe tempfile appendtext
'.\modify.exe' is not recognized as an internal or external command,
operable program or batch file.
scons: *** [file.out] Error 1
scons: building terminated because of errors.
```

What am I doing wrong here and why doesn't SCons recognize that `Command` needs `modify.exe` to be able to run? Thanks

**Comment by dmoody256** [comment]
I see a couple problems here:

1. you are using platform specific suffixes in your SConscript, now it is not portable. SCons will handle prefixes/suffixes for you automatically.
2. You have a directory which is potentially the same as a file target on linux (where program suffixes are nothing). In scons everything is a target and so there is no way differentiate a string target "modify" from the directory or file reference. You should change the name of your directory, something like `modify_app` or `modify_tool`. SCons could be made to work in this fashion but you will be hindering yourself because you will not be able to use the generic string target like `modify`.
3. The answer for the main question `why doesn't SCons recognize that Command needs modify.exe` is because it is not listed as a source. You can should add it as a source to your command.

Here are my suggested changes:

modify_app/SConscript.py
```py
env = Environment(tools=["mingw"])

modify_node = env.Program(target='#/modify', source="modify.cpp")
Return("modify_node")
```

SConstruct.py
```py
import os

env = Environment()

env['MODIFY'] = SConscript(["modify_app/SConscript"])

env['ENV']['PATH'] += f"{os.pathsep}." # so Command can execute 'modify'

env.Command(
    "file.out",
    ["$MODIFY", "file.in"],
    action = [
        Copy("tempfile", "${SOURCES[1]}"),
        "${SOURCES[0]} tempfile appendtext",
        Copy("$TARGET", "tempfile")
    ]
)
```
